src/NATS.Client.JetStream/Internal/NatsNatsJSSubBaseConsume.cs

```diff
@@ -0,0 +1,136 @@
+ *
+ *  - User Notifications CH:
+ *      Anything we want to let user know about the state of the
+ *      consumer, connection status, timeouts etc.
+ *
+ *  The main idea is to deliver user and control messages from the server
+ *  inbox subscription and internal control messages (e.g. heartbeat
+ *  timeouts) to a single 'controller' where all messages would be
+ *  processed in order and state managed in one place in a non-concurrent
+ *  manner so that races are avoided and it's easier to reason with state
+ *  changes.
+ *
+ *  User Notifications also have their own channel so they can be
+ *  prioritized and can run in their own Task where User error handler
+ *  will be dispatched.
+ *
+ *
+ *    NATS-SERVER
+ *        |                                              User
+ *        |             +--> [User Messages CH] -------> message loop
+ *        v           /                                  (await foreach)
+ *    [Sub CH] ---> Controller (with state)
+ *        ^           \                                  User error
+ *        |            +--> [User Notifications CH] ---> handler
+ *        |                                              (Action<>)
+ *        | Internal control msgs
+ *        | (e.g. heartbeat timeout)
+ *        |
+ *  Heartbeat Timer
+ *
+ */
+internal class NatsNatsJSSubBaseConsume<T> : NatsJSSubBase<T>, INatsJSSubConsume<T>
+{
+    private readonly Action<NatsJSNotification>? _errorHandler;
+    private readonly CancellationToken _cancellationToken;
+    private readonly Task _notifier;
+    private readonly Channel<NatsJSNotification> _notificationChannel;
+    private readonly Channel<NatsJSMsg<T?>> _userMessageChannel;
+
+    internal NatsNatsJSSubBaseConsume(
+        string stream,
+        string consumer,
+        NatsJSContext context,
+        ISubscriptionManager manager,
+        string subject,
+        NatsSubOpts? opts,
+        NatsJSSubState state,
+        INatsSerializer serializer,
+        Action<NatsJSNotification>? errorHandler = default,
+        CancellationToken cancellationToken = default)
+        : base(stream, consumer, context, manager, subject, opts, state, serializer, cancellationToken)
+    {
+        _errorHandler = errorHandler;
+        _cancellationToken = cancellationToken;
+
+        // User messages are buffered here separately to allow smoother flow while control loop
+        // pulls more data in the background. This also allows control messages to be dealt with
+        // in the same loop as the control messages to keep state updates consistent. This is as
+        // opposed to having a control and a message channel at the point of serializing the messages
+        // in NatsJSSub class.
+        _userMessageChannel = Channel.CreateBounded<NatsJSMsg<T?>>(new BoundedChannelOptions(1_000)
+        {
+            FullMode = BoundedChannelFullMode.Wait,
+            AllowSynchronousContinuations = false,
+        });
```

Comment:
should this take the options from `NatsSubOpts.ChannelOptions`?

Reply:
True 💯  I didn't think of that.